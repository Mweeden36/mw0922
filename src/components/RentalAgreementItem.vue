<template>
  <div class="rental-agreement-item" :class="itemType">
    <b class="rental-agreement__label">{{ label }}</b>
    <span class="rental-agreement__label">{{ displayValue }}</span>
  </div>
</template>

<script>
import tools from "@/data/tools.json";

export default {
  name: "RentalAgreementItem",

  props: {
    itemType: {
      type: String,
      required: true,
    },

    value: {
      required: true,
    },
  },

  data() {
    return {
      currencyFields: [
        "dailyCharge",
        "prediscountAmount",
        "discountAmount",
        "finalAmount",
      ],
      itemTypes: {
        toolCode: "Tool",
        checkoutDate: "Checkout Date",
        returnDate: "Return Date",
        discountPercent: "Discount Percent",
        chargeableDays: "Chargeable Days",
        dailyCharge: "Daily Charge",
        prediscountAmount: "Pre-discount Amount",
        discountAmount: "Discount Amount",
        finalAmount: "Final Amount",
      },
    };
  },

  computed: {
    isCurrencyField() {
      return this.currencyFields.indexOf(this.itemType) !== -1;
    },

    isPercentageField() {
      return this.itemType === "discountPercent";
    },

    label() {
      return this.itemTypes[this.itemType];
    },

    displayValue() {
      if (this.itemType === "toolCode") {
        const toolObj = tools.find((tool) => tool.code === this.value);
        return `${toolObj.brand} ${toolObj.type}`;
      }
      if (this.isCurrencyField) {
        return this.toCurrency(this.value);
      }
      if (this.isPercentageField) {
        return this.toPercentage(this.value);
      }
      return this.value;
    },
  },

  methods: {
    toPercentage(val) {
      return `${val}%`;
    },

    toCurrency(val) {
      // Convert to number for toLocaleString.
      const numericVal = Number(val);
      return `$${numericVal.toLocaleString(undefined, {
        minimumFractionDigits: 2,
      })}`;
    },
  },
};
</script>

<style lang="scss" scoped>
.rental-agreement-item {
  padding: 5px 0px;
  .rental-agreement__label {
    padding-right: 10px;
  }
}
</style>
