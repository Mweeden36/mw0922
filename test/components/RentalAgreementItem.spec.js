import RentalAgreementItem from "../../src/components/RentalAgreementItem.vue";
import { describe, it, expect } from "vitest";
import { mount } from "@vue/test-utils";

describe("RentalAgreementItem.vue", () => {
  it("should initialize when passing in the tool code", () => {
    const { vm } = mount(RentalAgreementItem, {
      props: {
        itemType: "toolCode",
        value: "JAKD",
      },
    });
    expect(vm.isCurrencyField).toEqual(false);
    expect(vm.isPercentageField).toEqual(false);
    expect(vm.label).toEqual("Tool");
    expect(vm.displayValue).toEqual("DeWalt jackhammer");
  });

  it("should initialize when passing in a percentage value", () => {
    const { vm } = mount(RentalAgreementItem, {
      props: {
        itemType: "discountPercent",
        value: 50,
      },
    });
    expect(vm.isCurrencyField).toEqual(false);
    expect(vm.isPercentageField).toEqual(true);
    expect(vm.label).toEqual("Discount Percent");
    expect(vm.displayValue).toEqual("50%");
  });

  it("should initialize when passing in a currency value", () => {
    const { vm } = mount(RentalAgreementItem, {
      props: {
        itemType: "finalAmount",
        value: 50124.125,
      },
    });
    expect(vm.isCurrencyField).toEqual(true);
    expect(vm.isPercentageField).toEqual(false);
    expect(vm.label).toEqual("Final Amount");
    expect(vm.displayValue).toEqual("$50,124.13");
  });

  it("should generate the correct html", () => {
    const wrapper = mount(RentalAgreementItem, {
      props: {
        itemType: "finalAmount",
        value: 50124.125,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
