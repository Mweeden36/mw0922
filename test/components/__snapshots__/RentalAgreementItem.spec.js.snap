// Vitest Snapshot v1

exports[`RentalAgreementItem.vue > should generate the correct html 1`] = `"<div class=\\"rental-agreement-item finalAmount\\" data-v-4db1ffff=\\"\\"><b class=\\"rental-agreement__label\\" data-v-4db1ffff=\\"\\">Final Amount</b><span class=\\"rental-agreement__label\\" data-v-4db1ffff=\\"\\">$50,124.13</span></div>"`;
